<h1><p style="text-align: center;">Datenanalyse mit Python <br>Projekt – 1</p></h1> <h1>- Polizeikontrollen im Straßenverkehr </h1>


# Analytische Auswertung / Analytical Exploration

In [1]:
import pandas as pd
from IPython.display import display

ri = pd.read_csv("RI_08Feb.csv")
pd.options.display.max_columns = None
display(ri.head())
import numpy as np

"""
ri = pd.read_csv('RI_cleaned.csv', nrows=50000)
drop_columns=["state", "county_name", "county_fips", "fine_grained_location"]
ri.drop(drop_columns, axis='columns', inplace=True)
ri.dropna(subset=['driver_gender'], inplace=True)
ri['is_arrested'] = ri.is_arrested.astype('bool')
combined = ri.stop_date.str.cat(ri.stop_time, sep=' ')
ri['stop_datetime'] = pd.to_datetime(combined)
ri.drop(["stop_date", "stop_time"], axis='columns', inplace=True)
ri.set_index('stop_datetime', inplace=True)
"""

/var/folders/bb/6m7zfmqn4dlckynjdkxpvrrr0000gn/T/ipykernel_9187/1714191368.py:4: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.
  ri = pd.read_csv("RI_08Feb.csv")


,stop_datetime,id,location_raw,police_department,driver_gender,driver_age_raw,driver_age,driver_race_raw,driver_race,violation_raw,violation,search_conducted,search_type_raw,search_type,contraband_found,stop_outcome,is_arrested,stop_duration,out_of_state,drugs_related_stop,district
0,2005-01-02 01:55:00,RI-2005-00001,Zone K1,600,M,1985.0,20.0,W,White,Speeding,Speeding,False,NaN,NaN,False,Citation,False,0-15 Min,False,False,Zone K1
1,2005-01-02 20:30:00,RI-2005-00002,Zone X4,500,M,1987.0,18.0,W,White,Speeding,Speeding,False,NaN,NaN,False,Citation,False,16-30 Min,False,False,Zone X4
2,2005-01-04 12:55:00,RI-2005-00004,Zone X4,500,M,1986.0,19.0,W,White,Equipment/Inspection Violation,Equipment,False,NaN,NaN,False,Citation,False,0-15 Min,False,False,Zone X4
3,2005-01-06 01:30:00,RI-2005-00005,Zone X4,500,M,1978.0,27.0,B,Black,Equipment/Inspection Violation,Equipment,False,NaN,NaN,False,Citation,False,0-15 Min,False,False,Zone X4
4,2005-01-12 08:05:00,RI-2005-00006,Zone X1,0,M,1973.0,32.0,B,Black,Call for Service,Other,False,NaN,NaN,False,Citation,False,30+ Min,True,False,Zone X1


'\nri = pd.read_csv(\'RI_cleaned.csv\', nrows=50000)\ndrop_columns=["state", "county_name", "county_fips", "fine_grained_location"]\nri.drop(drop_columns, axis=\'columns\', inplace=True)\nri.dropna(subset=[\'driver_gender\'], inplace=True)\nri[\'is_arrested\'] = ri.is_arrested.astype(\'bool\')\ncombined = ri.stop_date.str.cat(ri.stop_time, sep=\' \')\nri[\'stop_datetime\'] = pd.to_datetime(combined)\nri.drop(["stop_date", "stop_time"], axis=\'columns\', inplace=True)\nri.set_index(\'stop_datetime\', inplace=True)\n'

> Hat das **Geschlecht** eines Fahrers Einfluss auf das Verhalten der Polizei bei einer Verkehrskontrolle?  
**In diesem Kapitel** werden Sie diese Frage untersuchen und dabei Filterung, Gruppierung, Method Chaining, boolesche Logik, String-Methoden und mehr üben!


***

## 1. Untersuchung von Verkehrsverstößen

Bevor Sie die Verkehrsverstöße nach Geschlecht vergleichen, sollten Sie zunächst alle Verstöße untersuchen, um ein Grundverständnis für die Daten zu erhalten.

In dieser Übung zählen Sie die eindeutigen Werte in der Spalte `violation` und geben diese anschließend als Anteile (Proportionen) aus.

> Bevor Sie in diesem Abschnitt beginnen, **wiederholen Sie die Schritte aus dem vorherigen Kapitel zur Datenvorbereitung.**  
Fahren Sie in diesem Kapitel dort fort, wo Sie im vorherigen Kapitel aufgehört haben.

**ANWEISUNGEN**

*   Zählen Sie die eindeutigen Werte in der Spalte `violation`, um zu sehen, welche Verstöße von allen Fahrern begangen wurden.
*   Geben Sie die Häufigkeiten als Anteile des Gesamtwerts aus.

In [2]:
ri = pd.read_csv("RI_08Feb.csv")

/var/folders/bb/6m7zfmqn4dlckynjdkxpvrrr0000gn/T/ipykernel_9187/477615901.py:1: DtypeWarning: Columns (3) have mixed types. Specify dtype option on import or set low_memory=False.
  ri = pd.read_csv("RI_08Feb.csv")


In [3]:
ri.head()

,stop_datetime,id,location_raw,police_department,driver_gender,driver_age_raw,driver_age,driver_race_raw,driver_race,violation_raw,violation,search_conducted,search_type_raw,search_type,contraband_found,stop_outcome,is_arrested,stop_duration,out_of_state,drugs_related_stop,district
0,2005-01-02 01:55:00,RI-2005-00001,Zone K1,600,M,1985.0,20.0,W,White,Speeding,Speeding,False,NaN,NaN,False,Citation,False,0-15 Min,False,False,Zone K1
1,2005-01-02 20:30:00,RI-2005-00002,Zone X4,500,M,1987.0,18.0,W,White,Speeding,Speeding,False,NaN,NaN,False,Citation,False,16-30 Min,False,False,Zone X4
2,2005-01-04 12:55:00,RI-2005-00004,Zone X4,500,M,1986.0,19.0,W,White,Equipment/Inspection Violation,Equipment,False,NaN,NaN,False,Citation,False,0-15 Min,False,False,Zone X4
3,2005-01-06 01:30:00,RI-2005-00005,Zone X4,500,M,1978.0,27.0,B,Black,Equipment/Inspection Violation,Equipment,False,NaN,NaN,False,Citation,False,0-15 Min,False,False,Zone X4
4,2005-01-12 08:05:00,RI-2005-00006,Zone X1,0,M,1973.0,32.0,B,Black,Call for Service,Other,False,NaN,NaN,False,Citation,False,30+ Min,True,False,Zone X1


In [5]:
# Verstöße zählen
ri.violation.value_counts(dropna=False)

violation
Speeding               36111
Moving violation        6522
Equipment               3022
Registration/plates     1463
Other                    892
Name: count, dtype: int64

In [6]:
# Anteil der Verstöße am Gesamtwert (ergibt Werte zwischen 0 und 1)
ri.violation.value_counts(normalize=True)

violation
Speeding               0.752156
Moving violation       0.135847
Equipment              0.062945
Registration/plates    0.030473
Other                  0.018579
Name: proportion, dtype: float64

> Laut den normalisierten Werten wurde am häufigsten wegen Geschwindigkeitsüberschreitung angehalten, gefolgt von Verstößen beim Fahrverhalten („moving violation“).

***

## 2. Vergleich der Verstöße nach Geschlecht

Die Frage, die wir untersuchen, ist, ob männliche und weibliche Fahrer unterschiedliche Arten von Verkehrsverstößen begehen.

Zuerst erstellen Sie jeweils ein DataFrame für jedes Geschlecht und analysieren anschließend die Verstöße separat.

**ANWEISUNGEN**

*   Erstellen Sie ein DataFrame `female`, das nur Zeilen mit `driver_gender == 'F'` enthält.
*   Erstellen Sie ein DataFrame `male`, das nur Zeilen mit `driver_gender == 'M'` enthält.
*   Zählen Sie die Verstöße von Fahrerinnen und geben Sie sie als Anteile aus.
*   Zählen Sie die Verstöße von Fahrern und geben Sie sie ebenfalls als Anteile aus.


In [8]:
# Anzahl der Fahrer nach Geschlecht
ri.driver_gender.value_counts()

driver_gender
M    34701
F    13309
Name: count, dtype: int64

In [9]:
# DataFrame mit Fahrerinnen erstellen
female = ri[ri.driver_gender == "F"]

In [10]:
female.head(2)

,stop_datetime,id,location_raw,police_department,driver_gender,driver_age_raw,driver_age,driver_race_raw,driver_race,violation_raw,violation,search_conducted,search_type_raw,search_type,contraband_found,stop_outcome,is_arrested,stop_duration,out_of_state,drugs_related_stop,district
14,2005-02-24 01:20:00,RI-2005-00016,Zone X3,200,F,1983.0,22.0,W,White,Speeding,Speeding,False,NaN,NaN,False,Citation,False,0-15 Min,True,False,Zone X3
16,2005-03-14 10:00:00,RI-2005-00019,Zone K3,300,F,1984.0,21.0,W,White,Speeding,Speeding,False,NaN,NaN,False,Citation,False,0-15 Min,False,False,Zone K3


In [11]:
# DataFrame mit Fahrern erstellen
male = ri[ri.driver_gender == "M"]

In [12]:
male.head(2)

,stop_datetime,id,location_raw,police_department,driver_gender,driver_age_raw,driver_age,driver_race_raw,driver_race,violation_raw,violation,search_conducted,search_type_raw,search_type,contraband_found,stop_outcome,is_arrested,stop_duration,out_of_state,drugs_related_stop,district
0,2005-01-02 01:55:00,RI-2005-00001,Zone K1,600,M,1985.0,20.0,W,White,Speeding,Speeding,False,NaN,NaN,False,Citation,False,0-15 Min,False,False,Zone K1
1,2005-01-02 20:30:00,RI-2005-00002,Zone X4,500,M,1987.0,18.0,W,White,Speeding,Speeding,False,NaN,NaN,False,Citation,False,16-30 Min,False,False,Zone X4


In [14]:
# Anteil der Verstöße bei männlichen Fahrern
male.violation.value_counts() / male.violation.value_counts().sum()

## Das ist ein NumPy-Serie

violation
Speeding               0.729518
Moving violation       0.149967
Equipment              0.069595
Registration/plates    0.031584
Other                  0.019337
Name: count, dtype: float64

In [15]:
# Anteil der Verstöße bei Fahrerinnen
female.violation.value_counts(normalize=True)

violation
Speeding               0.811180
Moving violation       0.099031
Equipment              0.045608
Registration/plates    0.027575
Other                  0.016605
Name: proportion, dtype: float64

> Die normalisierten Werte zeigen, dass Frauen häufiger wegen Geschwindigkeitsüberschreitung angehalten werden, während Männer öfter wegen riskantem Fahrverhalten („moving violation“) auffallen.

## 3. Vergleich der Folgen von Geschwindigkeitsüberschreitungen nach Geschlecht

Viele Menschen glauben, dass das **Geschlecht** eines Fahrers Einfluss darauf hat, ob bei einem Stopp wegen **speeding** ein **"Citation"** (Ticket) oder nur eine **"Warning"** ausgesprochen wird.  
Gibt es dafür Hinweise im Datensatz?

Zuerst erstellen Sie zwei **DataFrames** mit Fahrern, die wegen **speeding** gestoppt wurden – einen für **females**, einen für **males**.

Anschließend berechnen Sie für jedes **gender**, wie viel Prozent der Stopps laut der Spalte **stop_outcome** zu einer **"Citation"** oder einer **"Warning"** führten.

**ANWEISUNGEN**

*   Erstellen Sie ein **DataFrame**, **female_and_speeding**, das nur female drivers enthält, die wegen **speeding** gestoppt wurden.
*   Erstellen Sie ein **DataFrame**, **male_and_speeding**, das nur male drivers enthält, die wegen **speeding** gestoppt wurden.
*   Zählen Sie die **stop outcomes** der female drivers und geben Sie sie als Anteile aus.
*   Zählen Sie die **stop outcomes** der male drivers und geben Sie sie ebenfalls als Anteile aus.


**Citation** = Verkehrsstrafe (Ticket)  
**Warning** = Verwarnung (keine Strafe)


In [18]:
# Nur 'female' drivers mit violation = "Speeding"
female_and_speeding = ri[(ri.driver_gender == "F") & (ri.violation == "Speeding")]

In [19]:
# Dasselbe DataFrame für 'male' drivers
male_and_speeding = ri[(ri.driver_gender == "M") & (ri.violation == "Speeding")]


In [22]:
male_and_speeding.head()

,stop_datetime,id,location_raw,police_department,driver_gender,driver_age_raw,driver_age,driver_race_raw,driver_race,violation_raw,violation,search_conducted,search_type_raw,search_type,contraband_found,stop_outcome,is_arrested,stop_duration,out_of_state,drugs_related_stop,district
0,2005-01-02 01:55:00,RI-2005-00001,Zone K1,600,M,1985.0,20.0,W,White,Speeding,Speeding,False,NaN,NaN,False,Citation,False,0-15 Min,False,False,Zone K1
1,2005-01-02 20:30:00,RI-2005-00002,Zone X4,500,M,1987.0,18.0,W,White,Speeding,Speeding,False,NaN,NaN,False,Citation,False,16-30 Min,False,False,Zone X4
5,2005-01-18 08:15:00,RI-2005-00007,Zone K3,300,M,1965.0,40.0,W,White,Speeding,Speeding,False,NaN,NaN,False,Citation,False,0-15 Min,True,False,Zone K3
7,2005-01-23 23:15:00,RI-2005-00009,Zone K3,300,M,1972.0,33.0,W,White,Speeding,Speeding,False,NaN,NaN,False,Citation,False,0-15 Min,True,False,Zone K3
8,2005-01-24 20:32:00,RI-2005-00010,Zone K1,600,M,1987.0,18.0,W,White,Speeding,Speeding,True,Probable Cause,Probable Cause,True,Citation,False,0-15 Min,True,True,Zone K1


In [23]:
female_and_speeding.head()

,stop_datetime,id,location_raw,police_department,driver_gender,driver_age_raw,driver_age,driver_race_raw,driver_race,violation_raw,violation,search_conducted,search_type_raw,search_type,contraband_found,stop_outcome,is_arrested,stop_duration,out_of_state,drugs_related_stop,district
14,2005-02-24 01:20:00,RI-2005-00016,Zone X3,200,F,1983.0,22.0,W,White,Speeding,Speeding,False,NaN,NaN,False,Citation,False,0-15 Min,True,False,Zone X3
16,2005-03-14 10:00:00,RI-2005-00019,Zone K3,300,F,1984.0,21.0,W,White,Speeding,Speeding,False,NaN,NaN,False,Citation,False,0-15 Min,False,False,Zone K3
23,2005-03-29 23:20:00,RI-2005-00026,Zone K3,300,F,1971.0,34.0,W,White,Speeding,Speeding,False,NaN,NaN,False,Citation,False,0-15 Min,True,False,Zone K3
32,2005-06-06 13:20:00,RI-2005-00035,Zone X4,500,F,1986.0,19.0,W,White,Speeding,Speeding,False,NaN,NaN,False,Citation,False,0-15 Min,False,False,Zone X4
34,2005-07-06 11:22:00,RI-2005-00038,Zone X1,0,F,1973.0,32.0,W,White,Speeding,Speeding,False,NaN,NaN,False,Citation,False,0-15 Min,False,False,Zone X1


In [24]:
# stop_outcome-Werte bei female_and_speeding als Anteile
female_and_speeding.stop_outcome.value_counts(normalize=True)

stop_outcome
Citation            0.973416
Warning             0.012968
Arrest Driver       0.007410
N/D                 0.003612
Arrest Passenger    0.002316
No Action           0.000278
Name: proportion, dtype: float64

In [25]:
# stop_outcome-Werte bei male_and_speeding als Anteile
male_and_speeding.stop_outcome.value_counts(normalize=True)

stop_outcome
Citation            0.957298
Arrest Driver       0.026230
Warning             0.010666
N/D                 0.003397
Arrest Passenger    0.002015
No Action           0.000395
Name: proportion, dtype: float64

> Sowohl female als auch male drivers erhalten bei speeding in etwa **zu 95 % eine Citation**.

> Allerdings zeigen die Daten, dass **male drivers häufiger verhaftet (Arrest)** wurden als female drivers.


In [26]:
# stop_outcome-Anteile für female drivers mit violation = "Speeding"
female[female['violation'] == 'Speeding'].stop_outcome.value_counts(normalize=True)

stop_outcome
Citation            0.973416
Warning             0.012968
Arrest Driver       0.007410
N/D                 0.003612
Arrest Passenger    0.002316
No Action           0.000278
Name: proportion, dtype: float64

In [27]:
# stop_outcome-Anteile für male drivers mit violation = "Speeding"
male[male["violation"] == "Speeding"].stop_outcome.value_counts(normalize=True)

stop_outcome
Citation            0.957298
Arrest Driver       0.026230
Warning             0.010666
N/D                 0.003397
Arrest Passenger    0.002015
No Action           0.000395
Name: proportion, dtype: float64

In [28]:
# Datentyp von search_conducted (ist ein boolescher Wert)
ri.search_conducted.dtype

dtype('bool')

In [31]:
# Anteil der Stopps mit bzw. ohne search_conducted - Normalisierung
ri.search_conducted.value_counts(normalize=True)

search_conducted
False    0.958092
True     0.041908
Name: proportion, dtype: float64

In [32]:
# Anteil der Stopps mit search_conducted = True - Normalisierung
ri.search_conducted.mean()

0.04190793584669861

!!! Die Logik von **normalize=True** ist identisch mit:

male.violation.value_counts(normalize=True)  
≙  
male.violation.value_counts() / male.violation.value_counts().sum()

oder z. B.:

for i in unique_vol:  
  print(i / unique_amount * 100)


In [33]:
# Anteil der female drivers, bei denen eine Durchsuchung durchgeführt wurde
female.search_conducted.mean()

0.017807498685100308

> Ergebnis: Etwa 1,7 % der female drivers wurden durchsucht.

In [34]:
# Anteil der male drivers mit durchgeführter Durchsuchung
male.search_conducted.mean()

0.05115126365234431

> Ergebnis: Etwa 5 % der male drivers wurden durchsucht.

In [35]:
# .unique gibt eine np-Serie mit den Werten "M" und "F" zurück
ri.driver_gender.unique()

# Beim groupby wird nach diesen Werten ("M" und "F") gruppiert

array(['M', 'F'], dtype=object)

In [36]:
# groupby alleine ist nicht aussagekräftig – Aggregationsfunktionen sind nötig
ri.groupby("driver_gender")  # axis=0, da auf Zeilen gearbeitet wird

In [37]:
# groupby ergibt ein DataFrame, daher kann man Spalten direkt ansprechen

# Durchschnitt von search_conducted je driver_gender (Anteil der Durchsuchungen)
ri.groupby("driver_gender").search_conducted.mean()

driver_gender
F    0.017807
M    0.051151
Name: search_conducted, dtype: float64

***

## 4. Berechnung der Durchsuchungsrate

Bei einer Verkehrskontrolle führt die Polizei manchmal eine Durchsuchung des Fahrzeugs durch.  
In dieser Übung berechnen Sie den Prozentsatz aller Stopps, bei denen eine Durchsuchung stattgefunden hat – auch bekannt als **search rate**.

**ANWEISUNGEN**

*   Überprüfen Sie den Datentyp von **search_conducted**, um sicherzustellen, dass es sich um eine **Boolean Series** handelt.
*   Berechnen Sie die **search rate**, indem Sie die Werte in der **Series** zählen und als Anteile ausdrücken.
*   Berechnen Sie die **search rate**, indem Sie den Mittelwert der **Series** nehmen. (Dieser sollte dem Anteil der **True**-Werte von oben entsprechen.)

In [38]:
ri.search_conducted.dtype

dtype('bool')

In [39]:
ri.search_conducted.value_counts(normalize=True)

search_conducted
False    0.958092
True     0.041908
Name: proportion, dtype: float64

In [40]:
ri.search_conducted.mean()

0.04190793584669861

***

## 5. Vergleich der Durchsuchungsraten nach Geschlecht

Sie vergleichen die Durchsuchungsraten von **female** und **male** drivers bei Verkehrskontrollen.  
Denken Sie daran: Die allgemeine **search rate** über alle Stopps liegt bei etwa **3.8 %**.

Zuerst filtern Sie das **DataFrame** nach Geschlecht und berechnen die **search rate** für jede Gruppe einzeln.  
Anschließend führen Sie dieselbe Berechnung für beide Gruppen gleichzeitig mit **.groupby()** durch.

**ANWEISUNGEN 1/3**

*   Filtern Sie das **DataFrame**, sodass nur **female** drivers enthalten sind,  
    und berechnen Sie anschließend die **search rate**, indem Sie den Mittelwert von **search_conducted** nehmen.

In [41]:
female.search_conducted.mean()

0.017807498685100308

**ANWEISUNGEN 2/3**

*   Filtern Sie das **DataFrame**, sodass nur **male** drivers enthalten sind,  
    und wiederholen Sie die Berechnung der **search rate**.

In [42]:
male.search_conducted.mean()

0.05115126365234431

**ANWEISUNGEN 3/3**

*   Verwenden Sie **groupby(driver_gender)**, um die **search rate** für beide Gruppen gleichzeitig zu berechnen.  
    (Das Ergebnis sollte mit den vorherigen Berechnungen übereinstimmen.)

In [43]:
ri.groupby("driver_gender")["search_conducted"].sum()

driver_gender
F     237
M    1775
Name: search_conducted, dtype: int64

***

## 6. Einbeziehen eines zweiten Faktors in die Analyse

Auch wenn die **search rate** bei **males** deutlich höher ist als bei **females**, könnte dieser Unterschied hauptsächlich auf einen zweiten Faktor zurückzuführen sein.

Zum Beispiel könnten Sie vermuten, dass die **search rate** je nach **violation**-Typ variiert – und dass der Unterschied zwischen males und females daran liegt, dass sie zu unterschiedlichen **violations** neigen.

Diese Hypothese können Sie testen, indem Sie die **search rate** für jede Kombination aus **driver_gender** und **violation** untersuchen.  
Wenn die Hypothese stimmt, sollten males und females bei jeder **violation** etwa gleich oft durchsucht werden.

Finden Sie unten heraus, ob das der Fall ist!


**ANWEISUNGEN 1/2**

*   Verwenden Sie **.groupby()**, um die **search rate** für jede Kombination aus **driver_gender** und **violation** zu berechnen.  
    Werden **males** und **females** bei jeder **violation** ungefähr gleich häufig durchsucht?


In [44]:
# Alle einzigartigen Werte in der Spalte violation (Verstoßarten)
ri.violation.unique()

array(['Speeding', 'Equipment', 'Other', 'Moving violation',
       'Registration/plates'], dtype=object)

In [45]:
# Gruppiert nach driver_gender und violation, dann Durchschnitt von search_conducted berechnen
ri.groupby(["driver_gender", "violation"]).search_conducted.mean()

driver_gender  violation          
F              Equipment              0.079077
               Moving violation       0.047800
               Other                  0.045249
               Registration/plates    0.114441
               Speeding               0.006854
M              Equipment              0.123395
               Moving violation       0.088778
               Other                  0.154993
               Registration/plates    0.171533
               Speeding               0.028560
Name: search_conducted, dtype: float64

**ANWEISUNGEN 2/2**

*   Ändern Sie die Reihenfolge und gruppieren Sie zuerst nach **violation**, dann nach **driver_gender**.  
    Die Ergebnisse lassen sich so möglicherweise leichter vergleichen.

In [46]:
# Wenn man die Gruppierreihenfolge ändert:
ri.groupby(["violation", "driver_gender"]).search_conducted.mean()

# Die erste Variante sah ordentlicher aus – diese hier gefällt mir nicht so 😄

violation            driver_gender
Equipment            F                0.079077
                     M                0.123395
Moving violation     F                0.047800
                     M                0.088778
Other                F                0.045249
                     M                0.154993
Registration/plates  F                0.114441
                     M                0.171533
Speeding             F                0.006854
                     M                0.028560
Name: search_conducted, dtype: float64

***

## 7. Zählen der Schutzdurchsuchungen

Bei einer Fahrzeugsuche kann der Polizeibeamte den Fahrer abtasten, um zu prüfen, ob er eine Waffe bei sich trägt.  
Dies wird als **"protective frisk"** bezeichnet.

Zuerst überprüfen Sie, wie oft **"Protective Frisk"** der einzige Suchtyp war.  
Dann verwenden Sie eine String-Methode, um alle Fälle zu finden, in denen der Fahrer durchsucht wurde.

**ANWEISUNGEN**

*   Zählen Sie die Werte in **search_type**, um zu sehen, wie oft **"Protective Frisk"** der einzige Suchtyp war.
*   Erstellen Sie eine neue Spalte **frisk**, die **True** ist, wenn **search_type** den String **"Protective Frisk"** enthält, sonst **False**.
*   Überprüfen Sie den Datentyp von **frisk**, um sicherzustellen, dass es sich um eine **Boolean Series** handelt.
*   Bilden Sie die Summe von **frisk**, um die Gesamtanzahl der durchgeführten Frisks zu zählen.


In [47]:
ri.search_type.value_counts()

search_type
Incident to Arrest                                          958
Probable Cause                                              244
Protective Frisk                                            204
Inventory                                                   117
Incident to Arrest,Inventory                                116
Incident to Arrest,Probable Cause                            76
Incident to Arrest,Protective Frisk                          63
Reasonable Suspicion                                         43
Probable Cause,Protective Frisk                              36
Incident to Arrest,Inventory,Protective Frisk                33
Inventory,Protective Frisk                                   23
Incident to Arrest,Probable Cause,Protective Frisk           20
Incident to Arrest,Inventory,Probable Cause                  19
Inventory,Probable Cause                                     16
Protective Frisk,Reasonable Suspicion                        16
Probable Cause,Reasonable Su

In [48]:
ri[ri.search_type == "Protective Frisk"].head(2)

,stop_datetime,id,location_raw,police_department,driver_gender,driver_age_raw,driver_age,driver_race_raw,driver_race,violation_raw,violation,search_conducted,search_type_raw,search_type,contraband_found,stop_outcome,is_arrested,stop_duration,out_of_state,drugs_related_stop,district
997,2005-10-06 21:35:00,RI-2005-01027,Zone X4,500,M,1966.0,39.0,W,White,Speeding,Speeding,True,Terry Frisk,Protective Frisk,False,Citation,False,0-15 Min,False,False,Zone X4
1088,2005-10-07 11:33:00,RI-2005-01121,Zone X4,900,M,1980.0,25.0,W,White,Equipment/Inspection Violation,Equipment,True,Terry Frisk,Protective Frisk,False,Arrest Driver,True,16-30 Min,False,False,Zone X4


In [49]:
# Neue Spalte "frisk": True, wenn "Protective Frisk" in search_type enthalten ist, sonst False
ri["frisk"] = ri.search_type.str.contains("Protective Frisk", na=False)

# Die Series muss genauso lang sein wie das DataFrame (z. B. 50.000 Zeilen)
# Da search_type Strings enthält, verwenden wir .str.contains()

In [50]:
ri.head()

,stop_datetime,id,location_raw,police_department,driver_gender,driver_age_raw,driver_age,driver_race_raw,driver_race,violation_raw,violation,search_conducted,search_type_raw,search_type,contraband_found,stop_outcome,is_arrested,stop_duration,out_of_state,drugs_related_stop,district,frisk
0,2005-01-02 01:55:00,RI-2005-00001,Zone K1,600,M,1985.0,20.0,W,White,Speeding,Speeding,False,NaN,NaN,False,Citation,False,0-15 Min,False,False,Zone K1,False
1,2005-01-02 20:30:00,RI-2005-00002,Zone X4,500,M,1987.0,18.0,W,White,Speeding,Speeding,False,NaN,NaN,False,Citation,False,16-30 Min,False,False,Zone X4,False
2,2005-01-04 12:55:00,RI-2005-00004,Zone X4,500,M,1986.0,19.0,W,White,Equipment/Inspection Violation,Equipment,False,NaN,NaN,False,Citation,False,0-15 Min,False,False,Zone X4,False
3,2005-01-06 01:30:00,RI-2005-00005,Zone X4,500,M,1978.0,27.0,B,Black,Equipment/Inspection Violation,Equipment,False,NaN,NaN,False,Citation,False,0-15 Min,False,False,Zone X4,False
4,2005-01-12 08:05:00,RI-2005-00006,Zone X1,0,M,1973.0,32.0,B,Black,Call for Service,Other,False,NaN,NaN,False,Citation,False,30+ Min,True,False,Zone X1,False


In [ ]:
# Anzahl der True-/False-Werte in der Spalte "frisk"
ri.frisk.value_counts()

frisk
False    47607
True       403
Name: count, dtype: int64

>  Es gibt 403 Zeilen, in denen "Protective Frisk" vorkommt

In [52]:
# Anteil der Zeilen, in denen search_type genau "Protective Frisk" ist (mit .isin())
ri.search_type.isin(["Protective Frisk"]).mean()

0.004249114767756718

In [53]:
# Zählt nur Zeilen, bei denen search_type exakt "Protective Frisk" ist
ri.search_type.isin(["Protective Frisk"]).sum()

# → Das reicht nicht aus, da andere Einträge wie "Protective Frisk, Vehicle Search" ignoriert werden

204

***

## 8. Vergleich der Frisk-Raten nach Geschlecht

Sie vergleichen die Raten, mit denen **female** und **male** drivers bei einer Durchsuchung gefilzt (**frisked**) werden.  
Werden **males** möglicherweise öfter gefilzt als **females**, weil sie von Polizeibeamten als risikoreicher eingeschätzt werden?

Bevor Sie Berechnungen durchführen, ist es wichtig, das **DataFrame** zu filtern, sodass nur die relevanten Daten enthalten sind – nämlich Stopps, bei denen eine **search** durchgeführt wurde.

**ANWEISUNGEN**

*   Erstellen Sie ein ``DataFrame`` namens ``searched``, das nur Zeilen enthält, in denen ``search_conducted`` ``True`` ist.
*   Berechnen Sie den Mittelwert der Spalte ``frisk``, um herauszufinden, bei welchem Prozentsatz der ``searches`` ein ``frisk`` durchgeführt wurde.
*   Berechnen Sie die ``frisk rate`` für jedes ``gender`` mit ``.groupby()``.

In [55]:
ri.search_conducted

# Das ist eine Series mit True- und False-Werten.
# Wenn man sie auf ``ri`` anwendet, erhält man ein ``DataFrame`` mit den Zeilen, bei denen der Wert ``True`` ist.

0        False
1        False
2        False
3        False
4        False
         ...  
48005    False
48006    False
48007    False
48008    False
48009    False
Name: search_conducted, Length: 48010, dtype: bool

In [56]:
# ``ri.search_conducted`` ist eine Series aus True/False mit gleicher Länge wie ``ri``
# Wird sie auf ``ri`` angewendet, entsteht ein neues ``DataFrame`` mit nur den ``True``-Zeilen
searched = ri[ri.search_conducted]

In [57]:
searched.head()

,stop_datetime,id,location_raw,police_department,driver_gender,driver_age_raw,driver_age,driver_race_raw,driver_race,violation_raw,violation,search_conducted,search_type_raw,search_type,contraband_found,stop_outcome,is_arrested,stop_duration,out_of_state,drugs_related_stop,district,frisk
8,2005-01-24 20:32:00,RI-2005-00010,Zone K1,600,M,1987.0,18.0,W,White,Speeding,Speeding,True,Probable Cause,Probable Cause,True,Citation,False,0-15 Min,True,True,Zone K1,False
9,2005-02-09 03:05:00,RI-2005-00011,Zone X4,500,M,1976.0,29.0,W,White,Registration Violation,Registration/plates,True,"Probable Cause,Terry Frisk","Probable Cause,Protective Frisk",False,Citation,False,0-15 Min,False,False,Zone X4,True
77,2005-08-28 01:00:00,RI-2005-00084,Zone X1,0,M,1979.0,26.0,W,White,Other Traffic Violation,Moving violation,True,"Incident to Arrest,Terry Frisk","Incident to Arrest,Protective Frisk",False,Arrest Driver,True,16-30 Min,True,False,Zone X1,True
83,2005-09-15 02:20:00,RI-2005-00094,Zone X4,500,M,1988.0,17.0,W,White,Other Traffic Violation,Moving violation,True,Incident to Arrest,Incident to Arrest,False,Arrest Driver,True,16-30 Min,False,False,Zone X4,False
104,2005-09-24 02:20:00,RI-2005-00115,Zone K3,300,M,1987.0,18.0,W,White,Other Traffic Violation,Moving violation,True,Incident to Arrest,Incident to Arrest,False,Arrest Driver,True,16-30 Min,False,False,Zone K3,False


In [58]:
searched.frisk.mean()

0.20029821073558648

In [ ]:
# Gruppieren nach ``driver_gender`` und Durchschnitt von ``frisk`` berechnen
# → zeigt, wie oft pro Geschlecht ein ``frisk`` durchgeführt wurde
searched.groupby("driver_gender").frisk.mean()



driver_gender
F    0.164557
M    0.205070
Name: frisk, dtype: float64

>  Etwa 16 % der Frauen und 20 % der Männer wurden gefilzt. 
→ Das ist ein Unterschied von 4 % (0,20 - 0,16 = 0,04). 
Das ist ein kleiner Unterschied, aber es gibt noch mehr Unterschiede.

### Fazit

Die Analyse zeigt, dass das Geschlecht der Fahrer einen Einfluss auf bestimmte polizeiliche Maßnahmen während Verkehrskontrollen haben kann.  
Obwohl die allgemeine Durchsuchungsrate niedrig ist, werden male drivers häufiger durchsucht und gefilzt als female drivers.  
Auch bei Geschwindigkeitsüberschreitungen erhalten beide Geschlechter ähnlich oft ein Ticket, jedoch gibt es leichte Unterschiede im Umgang mit „Citations“ und „Warnings“.  
Diese Erkenntnisse können helfen, mögliche Muster oder Ungleichbehandlungen datenbasiert zu identifizieren.